mlir-sycl/doc/design/SYCLGridRepresentation.md

```diff
@@ -0,0 +1,244 @@
+
+After the canonicalization pass, we might obtain a module in which a new
+operation implementing `SYCLMethodOpInterface` is introduced, but no
+corresponding definition is present in the module. To avoid this situation, the
+SYCL dialect provides a function definition register which corresponds to a map
+`(FunctionName, FunctionType) -> Definition`. Definitions are inserted using
+`SYCLDialect::addMethodDefinition(llvm::StringRef, mlir::func::FuncOp)` and
+retrieved using `SYCLDialect::lookupMethodDefinition(llvm::StringRef,
+mlir::FunctionType)`.
+
+In the codegen side, when code is generated for a function in the `sycl`
+namespace (an `mlir::func::FuncOp` is obtained), if this corresponds to an
+operation, it is inserted in the SYCL dialect. After this insertions takes
+place, only declarations can be overriden by functions of the same name.
+
+### Lowering
+
+All operations implementing `SYCLMethodOpInterface` are lowered to a `sycl.call`
+operation. In order to do so, the corresponding attributes of the operation are
+used to construct the `sycl.call`. Regarding the `MangledFunctionName`
+attribute, this was not needed in the original operation as the lowering process
+makes use of the definition register to obtain the function implementing the
+operation, insert it in the module if it is not present yet, and use the name of
+this function to build the attribute.
+
+In order to create a legal function call, the `FunctionArgTypes` attribute is
+used to perform the necessary transformations beforehand:
+
+1. Build a `memref.alloca()` of type `memref<1xTy>`, being `Ty` the type of the
+   first argument;
+2. Store the value there;
+3. If the argument type is a `memref` and the shapes differ, introduce a
+   `memref.cast` operation;
+4. If the memory spaces differ, use a `polygeist.memref2pointer`,
+   `llvm.addrspacecast` and `polygeist.pointer2memref` operation chain to cast
+   to the required one;
+5. If needed, cast to the base type using `sycl.cast` (or other operation we
+   might have in the future).
+
+#### Lowering example
+
+The following `sycl.nd_item.get_global_id` operation:
+
+```MLIR
+%1 = sycl.nd_item.get_global_id(%nd, %idx) {FunctionArgTypes = {!llvm.ptr<struct<...>, 4>}, FunctionName = @get_global_id, TypeName = @nd_item} : (!sycl_nd_item_1_, i32) -> i64
+```
+
+Would be lowered to:
+
+```MLIR
+%1 = memref.alloca() : memref<1x!sycl_nd_item_1_>
+memref.store %nd, %1[0] : memref<1x!sycl_nd_item_1_>
+%2 = polygeist.memref2pointer(%1) : (memref<1x!sycl_nd_item_1_>) -> !llvm.ptr<!sycl_nd_item_1_>
+%3 = llvm.addrspacecast %2 : !llvm.ptr<!sycl_nd_item_1_> to !llvm.ptr<!sycl_nd_item_1_, 4>
+%4 = polygeist.pointer2memref(%3) : (!llvm.ptr<!sycl_nd_item_1_, 4>) -> memref<1x!sycl_nd_item_1_, 4>
+%5 = sycl.cast(%4) : (memref<1x!sycl_nd_item_1_, 4>) -> !llvm.ptr<struct<...>, 4>
+%6 = sycl.call(%5, %idx) {FunctionName = @get_global_id, MangledFunctionName = <MFN>, TypeName = @nd_item} : (!llvm.ptr<struct<...>, 4>, i32) -> i64
```

Comment:
Can you explain how the compiler determines what mangled function name to use for the new operation created by canonicalization ?

Reply:
Using `SYCLDialect::lookupMethodDefinition(llvm::StringRef, mlir::FunctionType)`. This is what we have now (can be enabled by a flag, tested in `functions.cpp`).